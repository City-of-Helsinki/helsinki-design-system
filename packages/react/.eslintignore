**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
