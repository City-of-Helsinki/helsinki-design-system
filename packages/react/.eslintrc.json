{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
