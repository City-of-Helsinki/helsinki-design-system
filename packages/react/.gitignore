node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
