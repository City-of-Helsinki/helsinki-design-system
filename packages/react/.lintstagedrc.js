module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
