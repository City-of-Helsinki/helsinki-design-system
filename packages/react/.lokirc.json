{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
