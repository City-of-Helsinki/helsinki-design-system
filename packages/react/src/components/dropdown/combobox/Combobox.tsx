/* eslint-disable @typescript-eslint/ban-ts-ignore */
/* eslint-disable react/destructuring-assignment */
import React, { useRef, useState, KeyboardEvent, FocusEvent, FocusEventHandler, useMemo, useCallback } from 'react';
import { useCombobox, useMultipleSelection } from 'downshift';
import isEqual from 'lodash.isequal';
import uniqueId from 'lodash.uniqueid';
import { useVirtual } from 'react-virtual';

import 'hds-core';

import styles from './Combobox.module.scss';
import selectStyles from '../select/Select.module.scss';
import { FieldLabel } from '../../../internal/field-label/FieldLabel';
import classNames from '../../../utils/classNames';
import { IconAlertCircleFill, IconAngleDown, IconCrossCircle } from '../../../icons';
import { SelectedItems } from '../../../internal/selectedItems/SelectedItems';
import { multiSelectReducer, onMultiSelectStateChange, SelectCustomTheme, SelectProps } from '../select';
import { DROPDOWN_MENU_ITEM_HEIGHT, getIsInSelectedOptions } from '../dropdownUtils';
import { DropdownMenu } from '../../../internal/dropdownMenu/DropdownMenu';
import getIsElementFocused from '../../../utils/getIsElementFocused';
import getIsElementBlurred from '../../../utils/getIsElementBlurred';
import { useTheme } from '../../../hooks/useTheme';

type FilterFunction<OptionType> = (options: OptionType[], search: string) => OptionType[];

export type ComboboxProps<OptionType> = SelectProps<OptionType> & {
  /**
   * Custom aria-describedby added to the input element
   */
  'aria-describedby'?: string;
  /**
   * Prevents further propagation of the 'Escape' onKeyDown event when the menu is closed by pressing Esc.
   * Useful e.g. when the component is used inside a modal.
   */
  catchEscapeKey?: boolean;
  /**
   * If provided, this filter function will be used for filtering the
   * combobox suggestions. If this prop is not provided, the default
   * filter implementation is used. The default implementation assumes
   * that the `optionLabelField` prop points to a string value that it
   * can compare with the search value.
   */
  filter?: FilterFunction<OptionType>;
  /**
   * If `true`, displays a menu toggle button in the combobox.
   */
  showToggleButton?: boolean;
  /**
   * aria-label for the menu toggle button. The label for the combobox will be prepended to the given value.
   */
  toggleButtonAriaLabel: string;
};

function getDefaultFilter<OptionType>(labelField: string): FilterFunction<OptionType> {
  return (options: OptionType[], search: string) => {
    return options.filter((option) => {
      const label = option[labelField];
      const isLabelString = typeof label === 'string';

      if (!label) {
        // eslint-disable-next-line no-console
        console.warn(`Filtering failed because field ${labelField} could not be found from OptionType`);

        return false;
      }

      if (!isLabelString) {
        // eslint-disable-next-line no-console
        console.warn(`Filtering failed because field ${labelField} has a value that is not a string`);

        return false;
      }

      return label.toLowerCase().indexOf(search.toLowerCase()) > -1;
    });
  };
}

export const Combobox = <OptionType,>(props: ComboboxProps<OptionType>) => {
  // we can't destructure all the props. after destructuring, the link
  // between the multiselect prop and the value, onChange etc. props would vanish
  const {
    'aria-describedby': customAriaDescribedBy,
    catchEscapeKey,
    circularNavigation = false,
    className,
    clearable = props.multiselect,
    disabled = false,
    error,
    getA11ySelectionMessage = () => '',
    getA11yStatusMessage = () => '',
    helper,
    id = uniqueId('hds-combobox-') as string,
    invalid = false,
    isOptionDisabled,
    label,
    onBlur = () => null,
    onFocus = () => null,
    optionLabelField = 'label',
    options = [],
    placeholder,
    required,
    showToggleButton = true,
    style,
    theme,
    virtualized = false,
    visibleOptions = 5,
    filter: userLandFilter,
    toggleButtonAriaLabel,
    tooltipLabel,
    tooltipButtonLabel,
    tooltipText,
  } = props;

  // flag for whether the component is controlled
  const controlled = props.multiselect && props.value !== undefined;
  // custom theme class that is applied to the root element
  const customThemeClass = useTheme<SelectCustomTheme>(styles.root, theme);
  // selected items container ref
  const selectedItemsContainerRef = useRef<HTMLDivElement>();
  // combobox input ref
  const inputRef = useRef<HTMLInputElement>();
  // menu ref
  const menuRef = React.useRef<HTMLUListElement>();
  // toggle button ref
  const toggleButtonRef = React.useRef<HTMLButtonElement>(null);
  // whether active focus is within the dropdown
  const [hasFocus, setFocus] = useState<boolean>(false);
  // Tracks whether any combobox item is being clicked
  const [isClicking, setIsClicking] = useState<boolean>(false);
  // tracks current combobox search value
  const [search, setSearch] = useState<string>('');
  // memorise filtered items and only update them when any of the dependencies change
  const getFilteredItems = useMemo<OptionType[]>(() => {
    const filter = userLandFilter || getDefaultFilter(optionLabelField);
    return filter(options, search);
  }, [options, search, userLandFilter, optionLabelField]);
  // virtualize menu items to increase performance
  const virtualizer = useVirtual<HTMLUListElement>({
    size: getFilteredItems.length,
    parentRef: menuRef,
    estimateSize: useCallback(() => DROPDOWN_MENU_ITEM_HEIGHT, []),
    overscan: visibleOptions,
  });

  const focusInput = () => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
  };

  // init multi-select
  const {
    activeIndex,
    addSelectedItem,
    getDropdownProps,
    getSelectedItemProps,
    removeSelectedItem,
    reset,
    selectedItems,
    setActiveIndex,
    setSelectedItems: _setSelectedItems,
  } = useMultipleSelection<OptionType>({
    // sets focus on the first selected item when the dropdown is initialized
    defaultActiveIndex: 0,
    initialActiveIndex: 0,
    // set the default value(s) when the dropdown is initialized
    ...(props.multiselect && { initialSelectedItems: props.defaultValue ?? [] }),
    // set the selected items when the dropdown is controlled
    ...(props.multiselect && props.value !== undefined && { selectedItems: props.value ?? [] }),
    getA11yRemovalMessage: (props.multiselect && props.getA11yRemovalMessage) ?? (() => ''),
    onSelectedItemsChange: ({ selectedItems: _selectedItems }) =>
      props.multiselect && typeof props.onChange === 'function' && props.onChange(_selectedItems),
    onStateChange: (changes) =>
      onMultiSelectStateChange<OptionType>(changes, activeIndex, selectedItemsContainerRef.current),
    stateReducer: (state, actionAndChanges) => multiSelectReducer<OptionType>(state, actionAndChanges, controlled),
  });

  // init combobox
  const {
    getItemProps,
    getLabelProps,
    getMenuProps,
    getToggleButtonProps,
    highlightedIndex,
    isOpen,
    reset: resetCombobox,
    selectedItem,
    selectItem,
    closeMenu,
    setInputValue,
    getInputProps,
    getComboboxProps,
  } = useCombobox<OptionType>({
    circularNavigation,
    id,
    items: getFilteredItems,
    // set the default value when the dropdown is initialized
    ...(props.multiselect === false && { initialSelectedItem: props.defaultValue ?? null }),
    // a defined value indicates that the dropdown should be controlled
    // don't set selectedItem if it's not, so that downshift can handle the state
    ...(props.multiselect === false && props.value !== undefined && { selectedItem: props.value }),
    onInputValueChange: ({ inputValue }) => setSearch(inputValue),
    getA11ySelectionMessage,
    getA11yStatusMessage,
    itemToString: (item): string => (item ? item[optionLabelField] ?? '' : ''),
    onHighlightedIndexChange: ({ highlightedIndex: _highlightedIndex }) => virtualizer.scrollToIndex(_highlightedIndex),
    onSelectedItemChange: ({ selectedItem: _selectedItem }) =>
      props.multiselect === false && typeof props.onChange === 'function' && props.onChange(_selectedItem),
    onStateChange({ type, selectedItem: _selectedItem }) {
      const { InputBlur, InputKeyDownEnter, ItemClick } = useCombobox.stateChangeTypes;

      if (
        (type === InputBlur || type === InputKeyDownEnter || type === ItemClick) &&
        props.multiselect &&
        _selectedItem
      ) {
        getIsInSelectedOptions(selectedItems, _selectedItem)
          ? _setSelectedItems(selectedItems.filter((item) => !isEqual(item, _selectedItem)))
          : addSelectedItem(_selectedItem);
        selectItem(null);
      }
    },
    stateReducer(state, { type, changes }) {
      const { ItemClick, InputBlur } = useCombobox.stateChangeTypes;
      const { selectedItem: _selectedItem, inputValue } = changes;

      // clear the selected item if the input value doesn't match the selected item label
      if (!props.multiselect && _selectedItem && _selectedItem[optionLabelField] !== inputValue) {
        return {
          ...changes,
          selectedItem: null,
        };
      }

      // clear the input value on blur if there's no selected item
      if (type === InputBlur && !props.multiselect && !_selectedItem) {
        return {
          ...changes,
          inputValue: '',
        };
      }

      // prevent the menu from being closed when the user selects an item by clicking
      if (type === ItemClick && props.multiselect) {
        return {
          ...changes,
          isOpen: state.isOpen,
          highlightedIndex: state.highlightedIndex,
          // reset input value
          inputValue: '',
        };
      }

      return changes;
    },
  });

  const showClearButtonForSingleSelect = clearable && !props.multiselect && selectedItem;

  const setSelectedItems = (itemToBeSelected: OptionType) => {
    getIsInSelectedOptions(selectedItems, itemToBeSelected)
      ? _setSelectedItems(selectedItems.filter((item) => !isEqual(item, itemToBeSelected)))
      : addSelectedItem(itemToBeSelected);
  };

  const handleWrapperClick = (e) => {
    const selectedItemsContainerEl = selectedItemsContainerRef.current;
    const selectedItemClicked = selectedItemsContainerEl !== e.target && selectedItemsContainerEl?.contains(e.target);

    if (!selectedItemClicked) {
      focusInput();
    }
  };

  const ignoreFocusHandlerWhenClickingItem = (handler: FocusEventHandler<HTMLDivElement>) => (
    event: FocusEvent<HTMLDivElement>,
  ) => {
    if (!isClicking) {
      handler(event);
    }
  };

  const handleWrapperFocus = (e: FocusEvent<HTMLDivElement>) => {
    if (getIsElementFocused(e)) {
      setFocus(true);
      onFocus();
    }
  };

  const handleWrapperBlur = (e: FocusEvent<HTMLDivElement>) => {
    if (getIsElementBlurred(e)) {
      setFocus(false);
      onBlur();
    }
  };

  const handleMultiSelectInputKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    // prevent further propagation
    if (catchEscapeKey && e.key === 'Escape') {
      e.stopPropagation();
    }

    // Only select an item with 'Space' if an index is highlighted. Otherwise, it should have the default behavior, add a space into the input value.
    if (e.key === ' ' && highlightedIndex > -1) {
      // Prevent 'Space' from typing a space into the input.
      e.preventDefault();
      const highlightedItem = getFilteredItems[highlightedIndex];
      setSelectedItems(highlightedItem);
    }

    // If the menu is open, prevent the events for dropdown from firing.
    if (isOpen && (e.key === 'Backspace' || e.key === 'ArrowLeft')) {
      // @ts-ignore
      e.nativeEvent.preventDownshiftDefault = true;
    }

    // Tab keypress should close the menu without selecting an item.
    if (e.key === 'Tab' && highlightedIndex > -1 && isOpen) {
      // @ts-ignore
      closeMenu();
    }
  };

  if (!props.multiselect) {
    // we call the getDropdownProps getter function when multiselect isn't enabled
    // in order to suppress the "You forgot to call the ..." error message thrown by downshift.
    // we only need to apply the getter props to the toggle button when multiselect is enabled.
    getDropdownProps({}, { suppressRefError: true });
  }

  // Input should be shown when the combobox is open, or when it's
  // closed, but no items are selected. The input should always be
  // visible when multiselect mode is turned off.
  const isInputVisible = !props.multiselect || isOpen || (!isOpen && selectedItems.length === 0);

  // screen readers should read the labels in the following order:
  // field label > helper text > error text > toggle button label
  // helper and error texts should only be read if they have been defined
  // prettier-ignore
  const inputAriaLabel =
    `${getLabelProps().id}${error ? ` ${id}-error` : ''}${helper ? ` ${id}-helper` : ''} ${getInputProps().id}`;

  const renderInput = () => (
    <input
      {...getInputProps({
        ...(invalid && { 'aria-invalid': true }),
        ...(props.multiselect && {
          ...getDropdownProps({
            // Change Downshift's default behavior with space.
            // Instead of typing a space character into the
            // search input, it now selects an item without
            // closing the dropdown menu.

            // Our custom keyDown handler also blocks other
            // dropdown key events when the menu is open. This
            // would normally be done with the
            // 'preventKeyAction' setting, but it would also
            // block our custom handler from executing, which
            // would break the special behavior we have
            // implemented for space. We want to block other key
            // actions in order to ensure that dropdown and
            // input props don't conflict.
            onKeyDown: handleMultiSelectInputKeyDown,
            ref: inputRef,
          }),
        }),
        type: 'text',
        disabled,
        required,
        role: getComboboxProps().role,
        'aria-expanded': getComboboxProps()['aria-expanded'],
        'aria-haspopup': getComboboxProps()['aria-haspopup'],
        'aria-owns': getComboboxProps()['aria-owns'],
        'aria-labelledby': inputAriaLabel,
        'aria-describedby': customAriaDescribedBy,
      })}
      placeholder={placeholder}
      className={classNames(
        styles.input,
        !isInputVisible && styles.hidden,
        !showToggleButton && styles.noToggle,
        hasFocus && selectedItems.length > 0 && styles.adjustSpacing,
        props.icon && props.multiselect && styles.inputWithIcon,
        clearable && !props.multiselect && selectedItem && styles.withClearButton,
      )}
      autoCorrect="off"
      autoComplete="off"
    />
  );

  return (
    <div
      className={classNames(
        styles.root,
        invalid && styles.invalid,
        disabled && styles.disabled,
        isOpen && styles.open,
        props.multiselect && styles.multiselect,
        customThemeClass,
        className,
      )}
      style={style}
    >
      {/* LABEL */}
      {label && (
        <FieldLabel
          label={label}
          required={required}
          {...getLabelProps()}
          tooltipLabel={tooltipLabel}
          tooltipButtonLabel={tooltipButtonLabel}
          tooltipText={tooltipText}
        />
      )}
      {
        // This onClick function is used so that mouse users are able to
        // focus the Combobox without having to use the keyboard. The
        // design calls for the input to be visually hidden until the
        // user gives indication of wanting to access it.
        // Keyboard and screen reader users will move through the
        // selected items list and the clear button, after which they
        // will find the input. It's assumed that mouse users can
        // consume this information all at once and it's hence
        // convenient to offer direct access to the input by clicking.
        // In turn, providing this element as a focusable element would
        // add one more item keyboard/screen reader users would need to
        // move through.
      }
      {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */}
      <div
        // When a user clicks on a combobox item, the focus on the page
        // is momentarily lost. This will cause an event to fire which
        // has its 'relatedTarget' field set as 'null'. An event like
        // this looks exactly like a valid blur/focus event could look
        // like. We can not distinct this scenario in the blur/focus
        // handler.

        // If we allow blur/focus to fire, the input will briefly be
        // displayed as unfocused, which makes the selected items list
        // limit its visible options, which in turn changes the dropdown
        // position, which in turn makes selecting an item by click
        // almost impossible.

        // As a hackfix, we track whether any item is being clicked in
        // the components state. If an item is clicked, we do not call
        // the blur/focus handlers at all.
        onFocus={ignoreFocusHandlerWhenClickingItem(handleWrapperFocus)}
        onBlur={ignoreFocusHandlerWhenClickingItem(handleWrapperBlur)}
        onClick={handleWrapperClick}
        className={classNames(styles.wrapper, props.multiselect && props.icon && styles.wrapperWithMultiSelectAndIcon)}
        ref={getComboboxProps().ref}
      >
        {/* SELECTED ITEMS */}
        {props.multiselect && selectedItems.length > 0 && (
          <SelectedItems<OptionType>
            activeIndex={activeIndex}
            className={styles.selectedItems}
            clearable={clearable}
            clearButtonAriaLabel={props.clearButtonAriaLabel}
            dropdownId={id}
            getSelectedItemProps={getSelectedItemProps}
            hideItems={!hasFocus}
            onClear={() => {
              reset();
              setInputValue('');
              toggleButtonRef.current.focus();
            }}
            onRemove={removeSelectedItem}
            optionLabelField={optionLabelField}
            removeButtonAriaLabel={props.selectedItemRemoveButtonAriaLabel}
            selectedItems={selectedItems}
            selectedItemSrLabel={props.selectedItemSrLabel}
            selectedItemsContainerRef={selectedItemsContainerRef}
            setActiveIndex={setActiveIndex}
            toggleButtonHidden={!showToggleButton}
          />
        )}
        {props.multiselect === false && props.icon && (
          <span className={styles.icon} aria-hidden>
            {props.icon}
          </span>
        )}
        {/* FILTER INPUT */}
        {props.multiselect && props.icon ? (
          <div className={classNames(styles.multiselectIconAndInputWrapper, !isInputVisible && styles.hidden)}>
            <span
              className={classNames(styles.icon, hasFocus && selectedItems.length > 0 && styles.adjustSpacingForIcon)}
              aria-hidden
            >
              {props.icon}
            </span>
            {renderInput()}
          </div>
        ) : (
          renderInput()
        )}
        {/* TOGGLE BUTTON */}
        <button
          type="button"
          {...getToggleButtonProps({
            disabled,
            className: classNames(styles.button, !showToggleButton && styles.hidden),
            'aria-label': `${label}: ${toggleButtonAriaLabel}`,
            'aria-expanded': isOpen,
            ...(invalid && { 'aria-invalid': true }),
            ref: toggleButtonRef,
          })}
        >
          <IconAngleDown className={styles.angleIcon} aria-hidden />
        </button>
        {showClearButtonForSingleSelect && (
          <button
            type="button"
            className={classNames(selectStyles.clearButton, !showToggleButton && selectStyles.noToggle)}
            onClick={() => {
              resetCombobox();
              toggleButtonRef.current.focus();
            }}
            aria-label={props.clearButtonAriaLabel}
          >
            <IconCrossCircle />
          </button>
        )}
        {/* MENU */}
        <DropdownMenu<OptionType>
          getItemProps={(item, index, selected, optionDisabled, virtualRow) =>
            getItemProps({
              item,
              index,
              disabled: optionDisabled,
              className: classNames(
                styles.menuItem,
                highlightedIndex === index && styles.highlighted,
                selected && styles.selected,
                optionDisabled && styles.disabled,
                virtualized && styles.virtualized,
              ),
              onMouseDown: () => {
                setIsClicking(true);
              },
              // We can't use 'onMouseDown' because it is fired
              // before 'onClick' which is too soon for us. Using
              // 'onClick' creates a niche case where it's
              // possible that the user fails to complete their
              // click. In other words, they mouse down on a
              // different element than they mouse up on. In this
              // scenario, the blur/focus events will be ignored
              // until the next successful click.
              onClick: () => {
                setIsClicking(false);
              },
              // apply styles for virtualization to menu items
              ...(virtualRow && {
                style: {
                  transform: `translateY(${virtualRow.start}px`,
                },
                ref: virtualRow.measureRef,
              }),
            })
          }
          isOptionDisabled={isOptionDisabled}
          menuProps={getMenuProps({
            ...(props.multiselect && { 'aria-multiselectable': true }),
            ...(required && { 'aria-required': true }),
            style: { maxHeight: DROPDOWN_MENU_ITEM_HEIGHT * visibleOptions },
            ref: menuRef,
            onMouseLeave: (event) => {
              // prevent downshift from resetting highlighted index on mouseleave
              // @ts-ignore
              // eslint-disable-next-line no-param-reassign
              event.nativeEvent.preventDownshiftDefault = true;
            },
          })}
          menuStyles={styles}
          multiselect={props.multiselect}
          open={isOpen}
          optionLabelField={optionLabelField}
          options={getFilteredItems}
          selectedItem={selectedItem}
          selectedItems={selectedItems}
          virtualizer={virtualized && virtualizer}
          visibleOptions={visibleOptions}
        />
      </div>
      {/* INVALID TEXT */}
      {invalid && error && (
        <div id={`${id}-error`} className={styles.errorText} aria-hidden>
          <IconAlertCircleFill className={styles.invalidIcon} />
          {error}
        </div>
      )}
      {/* HELPER TEXT */}
      {helper && (
        <div id={`${id}-helper`} className={styles.helperText} aria-hidden>
          {helper}
        </div>
      )}
    </div>
  );
};
Combobox.defaultProps = {
  multiselect: false,
};
