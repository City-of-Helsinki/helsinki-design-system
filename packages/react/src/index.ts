export * from './icons';
export * from './components';
export * from './ssr';
