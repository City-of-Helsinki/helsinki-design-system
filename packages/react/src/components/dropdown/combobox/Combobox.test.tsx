import React from 'react';
import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { Combobox, ComboboxProps } from './Combobox';

const label = 'Combobox';

const options = [
  { label: 'Finland', value: 'FI' },
  { label: 'Sweden', value: 'SV' },
  { label: 'Norway', value: 'NO' },
  { label: 'Botswana', value: 'BW' },
  { label: 'Bolivia', value: 'BO' },
];

const multiWordOptions = [
  { label: 'Helsinki', value: 'helsinki' },
  { label: 'Helsinki east region', value: 'helsinki-east' },
  { label: 'Helsinki north region', value: 'helsinki-north' },
  { label: 'Helsinki northeast region', value: 'helsinki-northeast' },
  { label: 'Helsinki northwest region', value: 'helsinki-northwest' },
  { label: 'Helsinki west region', value: 'helsinki-west' },
];

const defaultProps: ComboboxProps<{ label: string; value: string }> = {
  label,
  options,
  clearButtonAriaLabel: 'Clear all selections',
  selectedItemRemoveButtonAriaLabel: 'Remove item {value}',
  selectedItemSrLabel: 'Selected item {value}',
  toggleButtonAriaLabel: 'Open the combobox',
};

const getWrapper = (props?: unknown) => render(<Combobox {...defaultProps} {...props} />);

describe('<Combobox />', () => {
  it('renders the component', () => {
    const { asFragment } = getWrapper();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = getWrapper();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('user should be able to search and choose an option ', async () => {
    const onChange = jest.fn();
    const { getAllByLabelText, getAllByRole, queryByDisplayValue } = getWrapper({ onChange });
    const input = getAllByLabelText(label)[0];
    userEvent.type(input, 'Fi');
    const visibleOptions = getAllByRole('option');
    // Ensure that options are filtered correctly
    expect(visibleOptions.length).toBe(1);

    // Choose one option
    userEvent.click(visibleOptions[0]);
    await waitFor(() => {
      // Ensure that chosen option is shown as value of input
      expect(queryByDisplayValue(options[0].label)).toBeDefined();
      // Ensure that chosen option has been sent with the onChange event
      expect(onChange).toHaveBeenCalledWith(options[0]);
    });
  });

  describe('in multi select mode', () => {
    it('user should be able to search and choose multiple options', async () => {
      const onChange = jest.fn();
      const { getAllByLabelText, getAllByRole, queryAllByText } = getWrapper({
        onChange,
        multiselect: true,
      });

      const input = getAllByLabelText(label)[0];

      // Search an option
      userEvent.type(input, 'Fi');
      const visibleOptions = getAllByRole('option');
      // Ensure that options are filtered correctly
      expect(visibleOptions.length).toBe(1);
      // Choose one option
      userEvent.click(visibleOptions[0]);
      await waitFor(() => {
        // Ensure that it's visible in selected items
        expect(queryAllByText(options[0].label).length).toEqual(2);
        // Ensure that it has been passed upwards with onChange
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith([options[0]]);
      });

      // Search another option
      userEvent.type(input, 'bo');
      expect(getAllByRole('option').length).toBe(2);
      userEvent.click(getAllByRole('option')[1]);
      await waitFor(() => {
        // Ensure that previous and current selection are visible in
        // selected items
        expect(queryAllByText(options[0].label).length).toEqual(2);
        expect(queryAllByText(options[4].label).length).toEqual(2);
        // Ensure that the current selection is passed correctly with
        // onChange
        expect(onChange).toHaveBeenCalledTimes(2);
        expect(onChange).toHaveBeenCalledWith([options[0], options[4]]);
      });
    });

    it('user should be able to close the menu with a tab keypress', () => {
      const onChange = jest.fn();
      const { getAllByLabelText, queryAllByRole, queryAllByText } = getWrapper({ onChange });
      const input = getAllByLabelText(label)[0];

      // Search and focus on first option
      userEvent.type(input, 'Fi');
      userEvent.type(input, '{arrowdown}');
      const visibleOptions = queryAllByRole('option');
      expect(visibleOptions.length).toBe(1);
      expect(visibleOptions[0].innerHTML).toEqual(options[0].label);
      expect(visibleOptions[0]).toHaveClass('highlighted'); // Prone to break if className changes

      // A tab keypress should close the menu without selecting an item
      userEvent.type(input, '{tab}');
      const visibleOptionsAfterClose = queryAllByRole('option');
      expect(visibleOptionsAfterClose.length).toBe(0);
      expect(queryAllByText(options[0].label).length).toEqual(0);
    });

    it('user should be able to search with multiple word term', () => {
      const onChange = jest.fn();
      const { getAllByLabelText, getAllByRole } = getWrapper({
        onChange,
        multiselect: true,
        options: multiWordOptions,
      });

      const input = getAllByLabelText(label)[0];

      // First result
      userEvent.type(input, 'Helsinki');
      const visibleOptions = getAllByRole('option');
      expect(visibleOptions.length).toBe(6);

      // The result after more specific search
      userEvent.type(input, '{space}northeast');
      const newVisibleOptions = getAllByRole('option');
      expect(newVisibleOptions.length).toBe(1);
      expect(newVisibleOptions[0]).toHaveTextContent('Helsinki northeast');
    });

    it('user should be able to search with a more specific term', () => {
      const onChange = jest.fn();
      const { getAllByLabelText, getAllByRole } = getWrapper({
        onChange,
        multiselect: true,
        options: multiWordOptions,
      });

      const input = getAllByLabelText(label)[0];
      userEvent.type(input, 'northeast');
      const visibleOptions = getAllByRole('option');
      expect(visibleOptions.length).toBe(1);
    });
  });
});
