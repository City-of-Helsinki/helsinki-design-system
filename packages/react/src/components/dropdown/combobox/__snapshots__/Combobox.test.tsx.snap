// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Combobox /> renders the component 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <label
      class="label "
      for="hds-combobox-1-input"
      id="hds-combobox-1-label"
    >
      Combobox
    </label>
    <div
      class="wrapper"
    >
      <input
        aria-autocomplete="list"
        aria-controls="hds-combobox-1-menu"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="hds-combobox-1-label hds-combobox-1-input"
        aria-owns="hds-combobox-1-menu"
        autocomplete="off"
        autocorrect="off"
        class="input"
        id="hds-combobox-1-input"
        role="combobox"
        type="text"
        value=""
      />
      <button
        aria-expanded="false"
        aria-label="Combobox: Open the combobox"
        class="button"
        id="hds-combobox-1-toggle-button"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="icon s angleIcon"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <ul
        aria-labelledby="hds-combobox-1-label"
        class="menu"
        id="hds-combobox-1-menu"
        role="listbox"
        style="max-height: 260px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
